L1Trigger/DTTriggerPhase2/src/MPQualityEnhancerFilterBayes.cc

```diff
@@ -0,0 +1,270 @@
+      ++SL1_shared;
+    }
+  }
+  if (mp.wi3 != -1 and mp.tdc3 != -1) {
+    ++SL1;
+    if (mp.wi3 == second_mp.wi3 and mp.tdc3 == second_mp.tdc3) {
+      ++SL1_shared;
+    }
+  }
+  if (mp.wi4 != -1 and mp.tdc4 != -1) {
+    ++SL1;
+    if (mp.wi4 == second_mp.wi4 and mp.tdc4 == second_mp.tdc4) {
+      ++SL1_shared;
+    }
+  }
+
+  if (mp.wi5 != -1 and mp.tdc5 != -1) {
+    ++SL3;
+    if (mp.wi5 == second_mp.wi5 and mp.tdc5 == second_mp.tdc5) {
+      ++SL3_shared;
+    }
+  }
+  if (mp.wi6 != -1 and mp.tdc6 != -1) {
+    ++SL3;
+    if (mp.wi6 == second_mp.wi6 and mp.tdc6 == second_mp.tdc6) {
+      ++SL3_shared;
+    }
+  }
+  if (mp.wi7 != -1 and mp.tdc7 != -1) {
+    ++SL3;
+    if (mp.wi7 == second_mp.wi7 and mp.tdc7 == second_mp.tdc7) {
+      ++SL3_shared;
+    }
+  }
+  if (mp.wi8 != -1 and mp.tdc8 != -1) {
+    ++SL3;
+    if (mp.wi8 == second_mp.wi8 and mp.tdc8 == second_mp.tdc8) {
+      ++SL3_shared;
+    }
+  }
+
+  // If the two mp share all hits in a SL, we consider that they share that SL
+  if (SL1_shared == SL1 || SL3_shared == SL3)
+    output = 1;
+
+  return output;
+}
+
+int MPQualityEnhancerFilterBayes::BX(metaPrimitive mp) {
+  int bx;
+  bx = (int)round(mp.t0 / (float)LHC_CLK_FREQ);
+  return bx;
+}
+
+// Is this really needed?
+int MPQualityEnhancerFilterBayes::rango(metaPrimitive mp) {
+  // Correlated
+  if (mp.quality > 5)
+    return 2;
+  // uncorrelated
+  else
+    return 1;
+}
+
+void MPQualityEnhancerFilterBayes::filterCousins(std::vector<metaPrimitive> &inMPaths,
+                                                 std::vector<metaPrimitive> &outMPaths) {
+  // At the beginning, we want to keep all mpaths
+  std::vector<bool> keep_this(inMPaths.size(), true);
+
+  // If we have just one mpath, save it
+  if (inMPaths.size() == 1) {
+    if (debug_) {
+      printmP(inMPaths[0]);
+    }
+    outMPaths.push_back(inMPaths[0]);
+  }
+  // More than one mpath
+  else if (inMPaths.size() > 1) {
+    for (int i = 0; i < int(inMPaths.size()); i++) {
+      if (debug_) {
+        printmP(inMPaths[i]);
+      }
+      // If we have already decided to reject the candidate, skip it
+      if (keep_this[i] == false)
+        continue;
+      for (int j = i + 1; j < int(inMPaths.size()); j++) {
+        // If we have already decided to reject the candidate, skip it
+        if (keep_this[j] == false)
+          continue;
+        // Case they are the same, keep the first one
+        if (areSame(inMPaths[i], inMPaths[j]) == true)
+          keep_this[i] = false;
+        // Case they are cousins, keep the best one
+        if (areCousins(inMPaths[i], inMPaths[j]) != 0) {
+          cout << "They are cousins" << endl;
+
+          // In case both are correlated, they have to share a full SL
+          if (inMPaths[i].quality > 5 && inMPaths[j].quality > 5 && shareSL(inMPaths[i], inMPaths[j]) == 0) {
+            cout << "But don't share a whole SL" << endl;
```

Comment:
Please don't use thread unsafe `cout` in production code: either replace with some LogMessage, or enclose it  a `if (debug_)` check (but perhaps they can be simply removed, as they looks like being temporary `cout`'s only meant to check the correctness of the developments)